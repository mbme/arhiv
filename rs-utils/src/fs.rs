use std::{
    env, fs,
    io::ErrorKind,
    os::unix::prelude::MetadataExt,
    path::{Path, PathBuf},
};

use anyhow::{anyhow, bail, ensure, Context, Result};

use crate::{bytes_to_hex_string, get_file_hash_blake3, get_string_hash_blake3};

pub fn path_exists(path: impl AsRef<str>) -> bool {
    fs::metadata(path.as_ref()).is_ok()
}

pub fn path_to_string(path: impl Into<PathBuf>) -> Result<String> {
    let result = path
        .into()
        .to_str()
        .context("Failed to convert file path to string")?
        .to_string();

    Ok(result)
}

/// This won't follow symlinks
pub fn file_exists(path: &str) -> Result<bool> {
    match fs::symlink_metadata(path) {
        Ok(metadata) if !metadata.is_file() => Err(anyhow!("path isn't a file: {}", path)),

        Ok(_) => Ok(true),

        Err(_) => Ok(false),
    }
}

/// This won't follow symlinks
pub fn dir_exists(path: &str) -> Result<bool> {
    match fs::symlink_metadata(path) {
        Ok(metadata) if !metadata.is_dir() => Err(anyhow!("path isn't a directory: {}", path)),

        Ok(_) => Ok(true),

        Err(_) => Ok(false),
    }
}

pub fn ensure_dir_exists(path: &str) -> Result<()> {
    ensure!(dir_exists(path)?, "dir doesn't exist {}", path);

    Ok(())
}

pub fn is_empty_dir(path: &str) -> Result<bool> {
    Ok(fs::read_dir(path)?.next().is_none())
}

pub fn ensure_file_exists(path: &str) -> Result<()> {
    ensure!(file_exists(path)?, "file doesn't exist {}", path);

    Ok(())
}

pub fn get_symlink_target_path(path: &str) -> Result<String> {
    let target_path = fs::read_link(path).context("failed to read a symlink")?;

    path_to_string(target_path)
}

/// check if path1 and path2 belong to the same filesystem or not
pub fn is_same_filesystem(path1: &str, path2: &str) -> Result<bool> {
    let meta1 = fs::metadata(path1)?;
    let meta2 = fs::metadata(path2)?;

    Ok(meta1.dev() == meta2.dev())
}

#[must_use]
pub fn get_file_name(path: &str) -> &str {
    Path::new(path)
        .file_name()
        .expect("file must have name")
        .to_str()
        .expect("file name must be valid string")
}

pub fn get_file_size(path: &str) -> Result<u64> {
    let metadata = fs::metadata(path).context("failed to read file metadata")?;

    Ok(metadata.len())
}

#[must_use]
pub fn get_file_extension(path: &str) -> Option<String> {
    let extension = Path::new(path).extension()?.to_str()?.to_string();

    Some(extension)
}

pub fn remove_file_extension(path: &str) -> Result<String> {
    let filename = Path::new(path)
        .file_stem()
        .context("failed to remove file extension")?;

    path_to_string(filename)
}

#[must_use]
pub fn get_home_dir() -> Option<String> {
    env::var_os("HOME").map(|path| {
        path.into_string()
            .expect("HOME env var must be a valid string")
    })
}

/// `$XDG_CONFIG_HOME` or `$HOME/.config`
#[must_use]
pub fn get_config_home() -> Option<String> {
    if let Some(path) = env::var_os("XDG_CONFIG_HOME") {
        return path
            .into_string()
            .expect("XDG_CONFIG_HOME env var must be a valid string")
            .into();
    }

    if let Some(path) = env::var_os("HOME") {
        return format!(
            "{}/.config",
            path.into_string()
                .expect("HOME env var must be a valid string")
        )
        .into();
    }

    None
}

/// `$XDG_DOWNLOAD_DIR` or `$HOME/Downloads`
#[must_use]
pub fn get_downloads_dir() -> Option<String> {
    if let Some(path) = env::var_os("XDG_DOWNLOAD_DIR") {
        return path
            .into_string()
            .expect("XDG_DOWNLOAD_DIR env var must be a valid string")
            .into();
    }

    if let Some(path) = env::var_os("HOME") {
        return format!(
            "{}/Downloads",
            path.into_string()
                .expect("HOME env var must be a valid string")
        )
        .into();
    }

    None
}

// recursively search from current dir upwards for {file_name}
pub fn locate_dominating_file<S: Into<String>>(file_name: S) -> Result<String> {
    let file_name = file_name.into();

    let mut dir = env::current_dir().context("must be able to get current dir")?;

    loop {
        let candidate = format!(
            "{}/{}",
            &dir.to_str().expect("must be able to serialize path"),
            file_name,
        );

        if file_exists(&candidate).unwrap_or(false) {
            return Ok(candidate);
        }

        if let Some(parent) = dir.parent() {
            dir = parent.to_path_buf();
        } else {
            bail!("Can't locate dominating file {}", file_name);
        }
    }
}

pub fn move_file(src: impl AsRef<str>, dest: impl AsRef<str>) -> Result<()> {
    let src = src.as_ref();
    let dest = dest.as_ref();

    match fs::rename(src, dest) {
        Err(err) if err.raw_os_error() == Some(18) => {
            // check for Invalid cross-device link (os error 18)
        }
        Err(err) => {
            return Err(err).context("failed to rename file");
        }
        Ok(_) => {
            return Ok(());
        }
    };

    // if error is due to src and dest being on different file systems
    // then copy src into dest, and remove src

    fs::copy(src, dest).context("failed to copy file data")?;

    if let Err(err) = fs::remove_file(src) {
        log::warn!("Failed to remove source file {}: {}", src, err);
    }

    Ok(())
}

pub fn into_absolute_path(path: impl AsRef<str>, canonicalize: bool) -> Result<String> {
    let path = shellexpand::full(path.as_ref()).context("failed to expand path")?;

    let path = path.as_ref();

    let mut path = if path.starts_with('/') {
        PathBuf::from(path)
    } else {
        current_dir_relpath(path)
    };

    // NOTE: path must exist
    if canonicalize {
        path = path
            .canonicalize()
            .with_context(|| format!("failed to canonicalize path {}", path.to_string_lossy()))?;
    }

    path_to_string(path)
}

#[must_use]
pub fn is_readable(metadata: &fs::Metadata) -> bool {
    let mode = metadata.mode();

    // TODO check also user / group (uid/gid)

    let user_has_read_access = mode & 0o400 != 0;
    let user_has_exec_access = mode & 0o100 != 0;

    if metadata.is_dir() {
        return user_has_read_access && user_has_exec_access;
    }

    user_has_read_access
}

pub fn get_media_type(file_path: impl AsRef<str>) -> Result<String> {
    let file_path = file_path.as_ref();

    if let Some(kind) = infer::get_from_path(file_path).context("failed to infer media type")? {
        return Ok(kind.mime_type().to_string());
    }

    Ok("application/octet-stream".to_string())
}

pub fn infer_extension_by_file_mime_type(file_path: &str) -> Result<Option<String>> {
    if let Some(kind) = infer::get_from_path(file_path).context("failed to infer mime type")? {
        return Ok(Some(kind.extension().to_string()));
    }

    Ok(None)
}

#[must_use]
pub fn get_extension_for_mime_string(mime: &str) -> Option<String> {
    let extensions = mime_guess::get_mime_extensions_str(mime).unwrap_or_default();

    Some((*extensions.first()?).to_string())
}

#[must_use]
pub fn get_mime_from_path(path: impl AsRef<str>) -> String {
    mime_guess::from_path(path.as_ref())
        .first_or_octet_stream()
        .to_string()
}

pub fn get_dir_checksum(path: impl AsRef<str>) -> Result<String> {
    let mut items: Vec<(String, String)> = Vec::new();

    for entry in fs::read_dir(path.as_ref())? {
        let entry = entry?;

        let name = path_to_string(entry.file_name())?;

        let path = path_to_string(entry.path())?;

        let hash = if fs::metadata(&path)?.is_dir() {
            get_dir_checksum(&path)?
        } else {
            bytes_to_hex_string(&get_file_hash_blake3(&path)?)
        };

        items.push((name, hash));
    }

    // sort by name
    items.sort_by(|a, b| a.0.cmp(&b.0));

    let result: String = items
        .into_iter()
        .flat_map(|item| vec![item.0, item.1])
        .collect();

    Ok(get_string_hash_blake3(&result))
}

pub fn create_dir_if_not_exist(dir_path: impl Into<PathBuf>) -> Result<()> {
    let dir_path = dir_path.into();

    if !dir_path.exists() {
        fs::create_dir(&dir_path).context(anyhow!("failed to create dir {:?}", dir_path))?;
    }

    Ok(())
}

pub fn create_file_if_not_exist(file_path: impl Into<PathBuf>) -> Result<()> {
    let file_path = file_path.into();

    if !file_path.exists() {
        fs::File::create(&file_path).context(anyhow!("failed to create file {:?}", file_path))?;
    }

    Ok(())
}

pub fn remove_file_if_exists(file_path: &str) -> Result<()> {
    match fs::remove_file(file_path) {
        Err(err) if err.kind() == ErrorKind::NotFound => {}
        Err(err) => {
            return Err(err).context("failed to remove file");
        }
        Ok(_) => {}
    }

    Ok(())
}

#[must_use]
pub fn workspace_relpath(subpath: &str) -> String {
    // Here CARGO_MANIFEST_DIR is /typed-v/rs-utils/
    format!("{}/../{}", env!("CARGO_MANIFEST_DIR"), subpath)
}

#[must_use]
pub fn current_dir_relpath(subpath: &str) -> PathBuf {
    let mut resource = env::current_dir().expect("invalid current directory");

    resource.push(subpath);

    resource
}

#[must_use]
pub fn is_image_filename(filename: impl AsRef<str>) -> bool {
    let ext = filename.as_ref().rsplit('.').next().unwrap_or_default();

    ext.eq_ignore_ascii_case("png")
        || ext.eq_ignore_ascii_case("jpg")
        || ext.eq_ignore_ascii_case("jpeg")
        || ext.eq_ignore_ascii_case("svg")
}

#[cfg(test)]
mod tests {
    use crate::workspace_relpath;

    use super::*;

    #[test]
    fn test_get_media_type() {
        assert_eq!(
            get_media_type(workspace_relpath("resources/text.txt")).unwrap(),
            "application/octet-stream"
        );
        assert_eq!(
            get_media_type(workspace_relpath("resources/k2.jpg")).unwrap(),
            "image/jpeg"
        );
        assert_eq!(
            get_media_type(workspace_relpath("resources/favicon-16x16.png")).unwrap(),
            "image/png"
        );
    }
}
