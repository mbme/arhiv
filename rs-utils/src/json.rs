use anyhow::*;
use serde_json::{Map, Value};

pub fn merge_json(o1: Value, o2: Value) -> Result<Value> {
    match (o1, o2) {
        (Value::Object(mut o1), Value::Object(mut o2)) => {
            let mut map = Map::new();

            map.append(&mut o1);
            map.append(&mut o2);

            Ok(Value::Object(map))
        }

        _ => bail!("failed to merge json values"),
    }
}
#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    #[test]
    fn test_merge_json() -> Result<()> {
        assert_eq!(
            merge_json(
                json!({
                    "test1": 1,
                }),
                json!({
                    "test2": 2,
                })
            )?,
            json!({
                "test1": 1,
                "test2": 2,
            })
        );

        assert_eq!(
            merge_json(
                json!({
                    "test1": 1,
                }),
                json!({
                    "test1": 2,
                })
            )?,
            json!({
                "test1": 2,
            })
        );

        assert!(merge_json(
            json!({
                "test1": 1,
            }),
            Value::Null,
        )
        .is_err());

        Ok(())
    }
}
