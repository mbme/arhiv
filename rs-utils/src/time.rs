use chrono::{DateTime, Local, Utc};

pub const MIN_TIMESTAMP: Timestamp = DateTime::<Utc>::MIN_UTC;

pub type Timestamp = DateTime<Utc>;

pub fn now() -> Timestamp {
    Utc::now()
}

pub fn format_time(timestamp: Timestamp, fmt: &str) -> String {
    timestamp.with_timezone(&Local).format(fmt).to_string()
}
