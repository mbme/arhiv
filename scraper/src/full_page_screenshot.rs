use anyhow::{Context, Result};
use fantoccini::{wd::WebDriverCompatibleCommand, Client};
use serde_json::{json, Value};

use rs_utils::decode_base64;

// TODO for Firefox / Geckodriver: let base64 = GET "/session/{sessionId}/moz/screenshot/full"

// https://chromedevtools.github.io/devtools-protocol/
#[derive(Debug)]
struct CDPCommand {
    command: String,
    params: Value,
}

impl CDPCommand {
    pub fn new(command: impl Into<String>, params: Value) -> Self {
        CDPCommand {
            command: command.into(),
            params,
        }
    }
}

// based on https://github.com/stevepryde/thirtyfour/blob/81bdd199e561f131dfab6d4a8a78a57cf7566917/thirtyfour/src/extensions/cdp/chromecommand.rs
impl WebDriverCompatibleCommand for CDPCommand {
    fn endpoint(
        &self,
        base_url: &url::Url,
        session_id: Option<&str>,
    ) -> Result<url::Url, url::ParseError> {
        let session_id = session_id.as_ref().expect("session_id must be present");

        base_url.join(&format!("session/{session_id}/goog/cdp/execute"))
    }

    fn method_and_body(&self, _request_url: &url::Url) -> (http::Method, Option<String>) {
        let body = json!({
            "cmd": self.command,
            "params": self.params,
        })
        .to_string();

        (http::Method::POST, Some(body))
    }
}

pub async fn cdp_capture_full_page_screenshot(client: &Client) -> Result<Vec<u8>> {
    let page_layout_metrics = client
        .issue_cmd(CDPCommand::new("Page.getLayoutMetrics", json!({})))
        .await
        .context("failed to run CDP command")?;

    let width = &page_layout_metrics["cssContentSize"]["width"];
    let height = &page_layout_metrics["cssContentSize"]["height"];

    let page_layout_metrics = client
        .issue_cmd(CDPCommand::new(
            "Page.captureScreenshot",
            json!({
                "captureBeyondViewport": true,
                "clip": {
                    "x": 0,
                    "y": 0,
                    "width": width,
                    "height": height,
                    "scale": 1,
                },
            }),
        ))
        .await
        .context("failed to run CDP command")?;

    let data = page_layout_metrics["data"]
        .as_str()
        .context("screenshot data is missing")?;

    decode_base64(data)
}
