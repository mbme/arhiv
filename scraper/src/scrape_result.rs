use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct FacebookPostListItem {
    permalink: String,
    date: String,
    #[serde(rename = "dateISO")]
    date_iso: Option<String>,
    content: String,
    images: Vec<String>,
    videos: Vec<String>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct FacebookMobilePostListItem {
    permalink: String,
    date: String,
    #[serde(rename = "dateISO")]
    date_iso: Option<String>,
    preview: String,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields, tag = "typeName")]
pub enum ScrapedData {
    FacebookPost {
        permalink: String,
        date: String,
        #[serde(rename = "dateISO")]
        date_iso: Option<String>,
        content: String,
        images: Vec<String>,
    },
    FacebookPostList {
        posts: Vec<FacebookPostListItem>,
    },
    FacebookMobilePost {
        permalink: String,
        date: String,
        #[serde(rename = "dateISO")]
        date_iso: Option<String>,
        content: String,
    },
    FacebookMobilePostList {
        posts: Vec<FacebookMobilePostListItem>,
    },
    #[serde(rename_all = "camelCase")]
    IMDBFilm {
        title: String,
        #[serde(rename = "coverURL")]
        cover_url: String,
        release_date: String,
        original_language: String,
        countries_of_origin: String,
        creators: String,
        cast: String,
        seasons: Option<u8>,
        episodes: Option<u8>,
        duration: String,
        description: String,
    },
    #[serde(rename_all = "camelCase")]
    MyAnimeListAnime {
        title: String,
        #[serde(rename = "coverURL")]
        cover_url: String,
        release_date: String,
        creators: String,
        duration: String,
        description: String,
    },
    #[serde(rename_all = "camelCase")]
    SteamGame {
        #[serde(rename = "coverURL")]
        cover_url: String,
        name: String,
        release_date: String,
        developers: String,
        description: String,
    },
    #[serde(rename_all = "camelCase")]
    YakabooBook {
        #[serde(rename = "coverURL")]
        cover_url: String,
        title: String,
        authors: String,
        publication_date: String,
        description: String,
        translators: String,
        publisher: String,
        pages: u32,
        language: String,
    },
    Image {
        #[serde(rename = "imageURL")]
        image_url: String,
    },
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ScrapeResult {
    pub url: String,
    pub original_url: Option<String>,
    pub scraper_name: Option<String>,
    pub data: Option<ScrapedData>,
    pub error: Option<String>,
}
