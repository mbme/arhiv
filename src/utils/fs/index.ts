export * from './utils'
export * from './lockfile'
export * from './transaction'
