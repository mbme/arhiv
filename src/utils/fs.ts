import fs from 'fs'
import os from 'os'
import path from 'path'
import {
  merge,
  consumeAsyncIterable,
} from './index'

interface IGetFilesOpts {
  skipDir?: string[]
  recursive?: boolean
  fileNames?: boolean
}
const GET_FILES_OPTS = {
  skipDir: ['.git', 'node_modules'],
  recursive: true,
  fileNames: false,
}

// list files except skip dirs
export async function* getFiles(
  rootPath: string,
  options: IGetFilesOpts = GET_FILES_OPTS,
): AsyncIterableIterator<string> {
  const processedOpts = merge(GET_FILES_OPTS, options)

  for (const fileName of await fs.promises.readdir(rootPath)) {
    const filePath = path.resolve(rootPath, fileName)

    if (await isDirectory(filePath)) {
      if (!processedOpts.recursive) {
        continue
      }

      // traverse subdir only if it shouldn't be skipped
      if (processedOpts.skipDir!.includes(fileName)) {
        continue
      }

      yield* getFiles(filePath, options)
    } else {
      yield processedOpts.fileNames ? fileName : filePath
    }
  }
}

// TODO use "recursive" option of rmsync instead
export function rmrfSync(dir: string) {
  for (const file of fs.readdirSync(dir)) {
    const filePath = path.join(dir, file)

    if (fs.statSync(filePath).isDirectory()) {
      rmrfSync(filePath)
    } else {
      fs.unlinkSync(filePath)
    }
  }

  fs.rmdirSync(dir)
}

export const isFile = (filePath: string) => fs.promises.lstat(filePath).then((stats) => stats.isFile())
export const isDirectory = (filePath: string) => fs.promises.lstat(filePath).then((stats) => stats.isDirectory())

export async function listFiles(dir: string) {
  return consumeAsyncIterable(getFiles(dir, {
    recursive: false,
    fileNames: true,
  }))
}

export const createTempDir = () => fs.promises.mkdtemp(path.join(os.tmpdir(), 'v-'))

export async function withTempFiles(files: string[], cb: (tempFiles: string[]) => void | Promise<void>) {
  if (!files.length) {
    return cb([])
  }

  let dir: string | undefined
  try {
    // create temp dir
    dir = await createTempDir()

    // write temp files
    const paths = files.map((_, i) => path.join(dir!, `temp-file-${i}`))
    await Promise.all(paths.map((filePath, i) => fs.promises.writeFile(filePath, files[i])))

    await Promise.resolve(cb(paths))
  } finally { // do cleanup in any case
    if (dir) rmrfSync(dir)
  }
}

export const readText = (name: string) => fs.promises.readFile(name, 'utf8')
export const readJSON = async <T>(name: string) => JSON.parse(await readText(name)) as T

export const writeText = (name: string, data: string) => fs.promises.writeFile(name, data, 'utf8')
export const writeJSON = (name: string, data: object) => writeText(name, JSON.stringify(data, undefined, 2))

export const getFileSize = (filePath: string) => fs.promises.stat(filePath).then(stats => stats.size)

export async function moveFile(file: string, dir: string): Promise<string> {
  const newFile = path.join(dir, path.basename(file))

  await fs.promises.rename(file, newFile)

  return newFile
}
