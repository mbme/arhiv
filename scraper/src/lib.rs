use std::time::Duration;

use anyhow::{ensure, Context, Result};
use cookie::Cookie;
use fantoccini::{Client, ClientBuilder};
use full_page_screenshot::cdp_capture_full_page_screenshot;
use serde_json::json;
use tokio::fs as tokio_fs;

use rs_utils::{into_absolute_path, log, path_exists, Chromedriver};

use crate::scrape_result::ScrapeResult;
pub use crate::scrape_result::ScrapedData;

mod full_page_screenshot;
mod scrape_result;

#[derive(Default)]
pub struct ScraperOptions {
    pub debug: bool,
    pub manual: bool,
    pub emulate_mobile: bool,
    pub screenshot_file: Option<String>, // .png
}

impl ScraperOptions {
    pub fn new() -> Self {
        Default::default()
    }

    pub async fn scrape(&self, url: &str) -> Result<Vec<ScrapeResult>> {
        let mut args = vec!["--incognito", "--no-sandbox", "--disable-dev-shm-usage"];

        let headless = !self.debug && !self.manual;
        if headless {
            args.push("--headless");
            args.push("window-size=1920,1080");
            args.push("--disable-gpu");

            args.push("user-agent=Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36");
        }

        if self.debug {
            args.push("--auto-open-devtools-for-tabs");
        }

        let capabilities = json!({
            "goog:chromeOptions": {
                "args": args,
                "mobileEmulation": if self.emulate_mobile {
                    json!({
                        "deviceName": "iPad Pro"
                    })
                } else {
                    json!({})
                },
            },
        });

        let mut chromedriver = Chromedriver::new()?;
        chromedriver.debug = self.debug;

        let mut chromedriver_process = chromedriver.spawn()?;

        chromedriver.wait_for_ready(10).await?;

        let client = ClientBuilder::rustls()
            .capabilities(
                capabilities
                    .as_object()
                    .context("must be an object")?
                    .clone(),
            )
            .connect(&chromedriver.get_url())
            .await
            .context("failed to connect to WebDriver")?;

        // TODO handle plain file downloads
        let result = self.run_browser_scraper(&client, url).await;

        if result.is_ok() {
            if let Err(err) = self.maybe_capture_sreenshot(&client).await {
                log::error!("Failed to save screenshot: {err}");
            };
        }

        if let Err(err) = client.close().await {
            log::error!("Failed to close webdriver client: {err}");
        }

        if let Err(err) = chromedriver_process.kill().await {
            log::error!("Failed to close webdriver: {err}");
        }

        result
    }

    async fn run_browser_scraper(&self, client: &Client, url: &str) -> Result<Vec<ScrapeResult>> {
        self.configure_timeouts(client).await?;

        client.goto(url).await.context("failed to open url")?;

        self.maybe_run_helpers(client, url)
            .await
            .context("failed to run scraper helpers")?;

        self.inject_browser_scraper(client).await?;

        let result = client
            .execute_async(
                r#"
                    const [url, manual, callback] = arguments;

                    window.originalURL = new URL(url);

                    window._doneCallback = () => {
                      callback(window._scraper.results);
                    };

                    if (manual) {
                      window._scraper.injectScraperUI();
                    } else {
                      try {
                        await window._scraper.scrape();
                      } finally {
                        window._doneCallback();
                      }
                    }
                "#,
                vec![json!(url), json!(self.manual)],
            )
            .await
            .context("failed to run scrape scenario")?;

        serde_json::from_value(result).context("failed to parse scrape results")
    }

    async fn maybe_run_helpers(&self, client: &Client, original_url: &str) -> Result<()> {
        let url = client.current_url().await?;

        // pass steam age check
        // https://store.steampowered.com/agecheck/app/814380/
        if url.host_str().unwrap_or_default() == "store.steampowered.com"
            && url.path().starts_with("/agecheck/")
        {
            let cookie = Cookie::build("birthtime", "628466401")
                .domain("store.steampowered.com")
                .path("/")
                .secure(false)
                .http_only(true)
                .same_site(cookie::SameSite::Strict)
                .finish();

            client
                .add_cookie(cookie)
                .await
                .context("failed to add steam age check cookie")?;

            client
                .goto(original_url)
                .await
                .context("failed to open original steam url")?;

            return Ok(());
        }

        Ok(())
    }

    async fn inject_browser_scraper(&self, client: &Client) -> Result<()> {
        client
            .execute(include_str!("../dist/browser-scraper.js"), vec![])
            .await
            .context("failed to inject browser-scraper.js")?;

        Ok(())
    }

    async fn maybe_capture_sreenshot(&self, client: &Client) -> Result<()> {
        let screenshot_file = if let Some(ref screenshot_file) = self.screenshot_file {
            screenshot_file
        } else {
            return Ok(());
        };

        let data = cdp_capture_full_page_screenshot(client)
            .await
            .context("failed to capture page screenshot")?;

        let file_path = into_absolute_path(screenshot_file, false)?;

        ensure!(
            !path_exists(&file_path),
            "Failed to save screenshot: {file_path} already exists"
        );

        tokio_fs::write(&file_path, data)
            .await
            .context("Failed to save screenshot")?;

        log::info!("Saved page screenshot into {file_path}");

        Ok(())
    }

    async fn configure_timeouts(&self, client: &Client) -> Result<()> {
        let mut timeouts = client
            .get_timeouts()
            .await
            .context("Failed to get session timeouts")?;

        if self.manual {
            timeouts.set_script(Some(Duration::from_secs(30 * 60)));
            timeouts.set_page_load(Some(Duration::from_secs(60)));
        } else {
            timeouts.set_script(Some(Duration::from_secs(60)));
            timeouts.set_page_load(Some(Duration::from_secs(30)));
        }

        client
            .update_timeouts(timeouts)
            .await
            .context("Failed to update session timeouts")?;

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use crate::{ScrapeResult, ScraperOptions};

    async fn scrape(url: &str) -> Result<Vec<ScrapeResult>> {
        let scraper = ScraperOptions::new();

        scraper.scrape(url).await
    }

    #[tokio::test]
    #[ignore]
    async fn test_scrape_facebook_post() -> Result<()> {
        let result = scrape("https://www.facebook.com/theprodigyofficial/posts/pfbid02XeNwZbYFN8TeXtYrgSCRLPciWpfNWEu3HaUartDe7X5HUH8XGWeXYbHz8wKdREdml").await?;

        insta::assert_json_snapshot!(result, {
            "[].data.permalink" => "[permalink]",
            "[].data.images[]" => "[permalink]"
        });

        Ok(())
    }

    #[tokio::test]
    #[ignore]
    async fn test_scrape_facebook_post_with_multiple_images() -> Result<()> {
        let result = scrape("https://www.facebook.com/theprodigyofficial/posts/pfbid0WVYZ4VTe9sddydcCNQGe7NUjajU92iVjM4TQYJgDpo14hy7zfHpQpdH5s2bWsoqul").await?;

        insta::assert_json_snapshot!(result, {
            "[].data.permalink" => "[permalink]",
            "[].data.images[]" => "[permalink]"
        });

        Ok(())
    }

    #[ignore]
    #[tokio::test]
    async fn test_scrape_yakaboo_book() -> Result<()> {
        let result =
            scrape("https://www.yakaboo.ua/ua/stories-of-your-life-and-others.html").await?;

        insta::assert_json_snapshot!(result, {
            "[].data.coverURL" => "[permalink]"
        });

        Ok(())
    }

    #[ignore]
    #[tokio::test]
    async fn test_scrape_steam_game() -> Result<()> {
        let result = scrape(
            "https://store.steampowered.com/app/814380/Sekiro_Shadows_Die_Twice__GOTY_Edition/",
        )
        .await?;

        insta::assert_json_snapshot!(result, {
            "[].data.coverURL" => "[permalink]",
        });

        Ok(())
    }

    #[ignore]
    #[tokio::test]
    async fn test_scrape_myanimelist_movie() -> Result<()> {
        let result = scrape("https://myanimelist.net/anime/523/Tonari_no_Totoro").await?;

        insta::assert_json_snapshot!(result, {
            "[].data.coverURL" => "[permalink]",
        });

        Ok(())
    }

    #[ignore]
    #[tokio::test]
    async fn test_scrape_myanimelist_series() -> Result<()> {
        let result = scrape("https://myanimelist.net/anime/16498/Shingeki_no_Kyojin").await?;

        insta::assert_json_snapshot!(result, {
            "[].data.coverURL" => "[permalink]",
        });

        Ok(())
    }

    #[ignore]
    #[tokio::test]
    async fn test_scrape_imdb_film() -> Result<()> {
        let result = scrape("https://www.imdb.com/title/tt0133093/").await?;

        insta::assert_json_snapshot!(result, {
            "[].data.coverURL" => "[permalink]",
        });

        Ok(())
    }

    #[ignore]
    #[tokio::test]
    async fn test_scrape_imdb_series() -> Result<()> {
        let result = scrape("https://www.imdb.com/title/tt0098936/").await?;

        insta::assert_json_snapshot!(result, {
            "[].data.coverURL" => "[permalink]",
        });

        Ok(())
    }

    #[ignore]
    #[tokio::test]
    async fn test_scrape_imdb_mini_series() -> Result<()> {
        let result = scrape("https://www.imdb.com/title/tt8134186/").await?;

        insta::assert_json_snapshot!(result, {
            "[].data.coverURL" => "[permalink]",
        });

        Ok(())
    }

    #[ignore]
    #[tokio::test]
    async fn test_scrape_image() -> Result<()> {
        let result =
            scrape("https://upload.wikimedia.org/wikipedia/en/7/7d/Lenna_%28test_image%29.png")
                .await?;

        insta::assert_json_snapshot!(result);

        Ok(())
    }
}
