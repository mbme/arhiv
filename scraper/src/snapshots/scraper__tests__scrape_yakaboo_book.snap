---
source: scraper/src/lib.rs
assertion_line: 257
expression: result
---
[
  {
    "url": "https://old.yakaboo.ua/ua/stories-of-your-life-and-others.html",
    "originalUrl": "https://www.yakaboo.ua/ua/stories-of-your-life-and-others.html",
    "scraperName": "YakabooBookScraper",
    "data": {
      "typeName": "YakabooBook",
      "coverURL": "[permalink]",
      "title": "Stories of Your Life and Others",
      "authors": "Тед Чан",
      "publicationDate": "2020",
      "description": "Includes 'Story of Your Life' the basis for the major motion picture Arrival, starring Amy Adams, Forest Whitaker, Jeremy Renner, and directed by Denis Villeneuve.\n\nWith his masterful first collection, multiple-award-winning author Ted Chiang deftly blends human emotion and scientific rationalism in eight remarkably diverse stories, all told in his trademark precise and evocative prose.\n\nFrom a soaring Babylonian tower that connects a flat Earth with the firmament above, to a world where angelic visitations are a wondrous and terrifying part of everyday life; from a neural modification that eliminates the appeal of physical beauty, to an alien language that challenges our very perception of time and reality. . . Chiang's rigorously imagined fantasia invites us to question our understanding of the universe and our place in it.",
      "translators": "",
      "publisher": "Picador",
      "pages": 340,
      "language": "English"
    },
    "error": null
  }
]
